aiida/cmdline/utils/migration/v03_to_v04.py

```diff
@@ -0,0 +1,528 @@
+
+        elif link['type'] == 'calllink':
+            # rename `calllink` to `call_calc` if the target node is a calculation type node
+            if uuid_node_type_mapping[inp_uuid].startswith('node.process.calculation'):
+                link['type'] = 'call_calc'
+            # rename `calllink` to `call_work` if the target node is a workflow type node
+            elif uuid_node_type_mapping[inp_uuid].startswith('node.process.workflow'):
+                link['type'] = 'call_work'
+
+
+def migration_dbgroup_name_to_label_type_to_type_string(metadata, data):
+    """Apply migrations: 0021 - REV. 1.0.21
+    Rename dbgroup fields:
+    name -> label
+    type -> type_string
+    """
+    # For data.json
+    if 'Group' in data['export_data']:
+        for content in data['export_data']['Group'].values():
+            if 'name' in content:
+                content['label'] = content.pop('name')
+            if 'type' in content:
+                content['type_string'] = content.pop('type')
+    # For metadata.json
+    metadata_group = metadata['all_fields_info']['Group']
+    if 'name' in metadata_group:
+        metadata_group['label'] = metadata_group.pop('name')
+    if 'type' in metadata_group:
+        metadata_group['type_string'] = metadata_group.pop('type')
+
+
+def migration_dbgroup_type_string_change_content(data):
+    """Apply migrations: 0022 - REV. 1.0.22
+    Change type_string according to the following rule:
+    '' -> 'user'
+    'data.upf.family' -> 'data.upf'
+    'aiida.import' -> 'auto.import'
+    'autogroup.run' -> 'auto.run'
+    """
+    for content in data['export_data']['Group'].values():
+        if content['type_string'] == '':
+            content['type_string'] = 'user'
+        elif content['type_string'] == 'data.upf.family':
+            content['type_string'] = 'data.upf'
+        elif content['type_string'] == 'aiida.import':
+            content['type_string'] = 'auto.import'
+        elif content['type_string'] == 'autogroup.run':
+            content['type_string'] = 'auto.run'
+
+
+def migration_calc_job_option_attribute_keys(data):
+    """Apply migrations: 0023 - REV. 1.0.23
+    `custom_environment_variables` -> `environment_variables`
+    `jobresource_params` -> `resources`
+    `_process_label` -> `process_label`
+    `parser` -> `parser_name`
+    """
+
+    # Helper function
+    def _migration_calc_job_option_attribute_keys(attr_id, content):
+        """Apply migration 0023 - REV. 1.0.23 for both `node_attributes*` dicts in `data.json`"""
+        if data['export_data']['Node'][attr_id]['type'] == 'node.process.calculation.calcjob.CalcJobNode.':
+            for key in content:
+                # custom_environment_variables -> environment_variables
+                if key == 'custom_environment_variables':
+                    content['environment_variables'] = content.pop('custom_environment_variables')
+                elif key.startswith(r'custom\_environment\_variables'):
```

Comment:
These comparisons should not be "regex" strings but normal strings. This means you should also remove the back slashes as the underscores have no special meaning, i.e. `key.startswith('custom_environment_variables')` . Same for all below. For reference, these underscores were necessary in the SQL because there we are actually doing a like comparison and underscore is a special character. However, here in python we are doing simple string comparisons.

Actually, the above is even superfluous because this whole `elif` block is not necessary and can be remove. In the SQL it was necessary because due to the EAV schema in Django, the nested elements in this dictionary would have a key `custom_environmnet_variables.some_key` but in the archive it is just a dictionary `{'custom_environment_variables': {'some_key': 1}}`

Reply:
done